### PR TITLE
fix: throw error when target has DND on

### LINKED ISSUE #11282: Notify about DND status when trying to resend call notification
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Have a user that is on DND
2. Send them the call notification again

### Expected behaviour
Either it works or I know it didn't

### Actual behaviour
Moderator is not notified that no push notification was send.

### PR DESCRIPTION
**Pull request by skalidindi53**
### ☑️ Resolves

* Fix #11282 

### 🚧 Tasks

- Throws a new error if target is on DND

### CODE REVIEW

File: lib/Service/ParticipantService.php [hunk 1]
```diff
@@ -1181,6 +1184,16 @@ public function sendCallNotificationForAttendee(Room $room, Participant $current
 			throw new DoesNotExistException('Room mismatch');
 		}
 
+		$userStatus = $this->userStatusManager->getUserStatuses(array($attendee->getActorId()));
+
+		if (empty($userStatus)) {
+			throw new \InvalidArgumentException('User not found');
+		}
+
+		if ($userStatus[0]->getStatus() === IUserStatus::DND){
```
Comment (nickvergessen) [inline lib/Service/ParticipantService.php:1193]: The user having no status is default and okay when they are "offline".
So it should not fail in this case and just continue without checking the user status.
So let's combine both checks to:
```suggestion
		if (isset($userStatus[0]) && $userStatus[0]->getStatus() === IUserStatus::DND) {
```

File: lib/Service/ParticipantService.php [hunk 2]
```diff
@@ -1181,6 +1184,16 @@ public function sendCallNotificationForAttendee(Room $room, Participant $current
 			throw new DoesNotExistException('Room mismatch');
 		}
 
+		$userStatus = $this->userStatusManager->getUserStatuses(array($attendee->getActorId()));
```
Comment (nickvergessen) [inline lib/Service/ParticipantService.php:1187]: *new array syntax*
```suggestion
		$userStatus = $this->userStatusManager->getUserStatuses([$attendee->getActorId()]);
```

File: lib/Service/ParticipantService.php [hunk 3]
```diff
@@ -1181,6 +1184,16 @@ public function sendCallNotificationForAttendee(Room $room, Participant $current
...
+
+		if (empty($userStatus)) {
+			throw new \InvalidArgumentException('User not found');
+		}
+
+		if ($userStatus[0]->getStatus() === IUserStatus::DND){
+			throw new \InvalidArgumentException('User is on do not disturb mode, failed to send notification');
```
Comment (nickvergessen) [inline lib/Service/ParticipantService.php:1194]: I would throw:
```suggestion
			throw new \InvalidArgumentException('status');
```

So you have the reason in the correct format available as `$e->getMessage()` for the last part of https://github.com/nextcloud/spreed/issues/11282#issuecomment-2089819787

Comment (skalidindi53) [reply lib/Service/ParticipantService.php]: Thank you for the help, just updated the code with the suggested changes. Also, I'm a little confused with which controller should I add the `return new DataResponse(['error' => 'status'], Http::STATUS_BAD_REQUEST);`. Is it the CallController.php?

File: lib/Service/ParticipantService.php [hunk 4]
```diff
@@ -1181,6 +1184,11 @@ public function sendCallNotificationForAttendee(Room $room, Participant $current
 			throw new DoesNotExistException('Room mismatch');
 		}
 
+		$userStatus = $this->userStatusManager->getUserStatuses([$attendee->getActorId()]);
+		if (isset($userStatus[0]) && $userStatus[0]->getStatus() === IUserStatus::DND) {
```
Comment (nickvergessen) [inline lib/Service/ParticipantService.php:1188]: You can't know that, but the docs of the user status manager are lying. It's actually a map (userid being the key) not a list:
```suggestion
		if (isset($userStatus[$attendee->getActorId()]) && $userStatus[$attendee->getActorId()]->getStatus() === IUserStatus::DND) {
```

Comment (nickvergessen) [reply lib/Service/ParticipantService.php]: Docs fixed in https://github.com/nextcloud/server/pull/45219

**Comment by skalidindi53** [comment]
> Maybe you can also duplicate the test in
> 
> https://github.com/nextcloud/spreed/blob/0d1fc13e36e40a6cd8dfa80915ea77f06b835f80/tests/integration/features/callapi/notifications.feature#L63-L78
> 
> add a check for the the expected status code (and message) when the user status was set to `dnd` using
> 
> https://github.com/nextcloud/spreed/blob/e5327997bc3c54924bd8d3a157930dd517eb3cd2/tests/integration/features/bootstrap/FeatureContext.php#L4315

Added the test case as well

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**